File: 55. Jump Game.py
class Solution:
    def canJump(self, nums: List[int]) -> bool:
        if not nums:
            return False
        if len(nums) == 1:
                return True
        goal = len(nums) - 1 
        for i in range(len(nums)-2,-1,-1):
            if i + nums[i] >= goal:
                goal = i
        return not goal
        
class Solution:
    def canJump(self, nums: List[int]) -> bool:
        if not nums:
            return False
        if len(nums) == 1:
                return True
        m = 0
        for i in range(len(nums)):
            if i > m:
                return False
            m = max(m, i+nums[i])
        return True

File: 55. Jump Game.java
class Solution {
    public boolean canJump(int[] nums) {
        int further=0;
        for (int i=0; i<nums.length; i++) {
            if (further >= i) {
                further = Math.max(further, i+nums[i]);
            }
        }
        return further >= nums.length-1;
    }
}